core/templates/dev/head/pages/exploration_player/ImagePreloaderService.js

```diff
@@ -0,0 +1,197 @@
+     * Initiates the image preloader beginning from the sourceStateName.
+     * @param {string} sourceStateName - The name of the state from which
+     *                                   preloader should start.
+     */
+    var _kickOffImagePreloader = function(sourceStateName) {
+      _filenamesOfImageToBeDownloaded = (
+        _getImageFilenamesInBfsOrder(sourceStateName));
+      while (_filenamesOfImageCurrentlyDownloading.length <
+        MAX_NUM_IMAGE_FILES_TO_DOWNLOAD_SIMULTANEOUSLY &&
+          _filenamesOfImageToBeDownloaded.length > 0) {
+        var imageFilename = _filenamesOfImageToBeDownloaded.shift();
+        _filenamesOfImageCurrentlyDownloading.push(imageFilename);
+        _loadImage(imageFilename);
+      }
+    };
+
+    /**
+     * Cancels the preloading of the images that are being downloaded.
+     */
+    var _cancelPreloading = function() {
+      AssetsBackendApiService.abortAllCurrentImageDownloads();
+      _filenamesOfImageCurrentlyDownloading = [];
+    };
+
+    /**
+     * Called when the state changes.
+     * @param {string} stateName - The name of the state the user shifts to.
+     */
+    var _onStateChange = function(stateName) {
+      if (stateName !== _exploration.getInitialState().name) {
+        _imageLoadedCallback = {};
+        var imageFilenamesInState = [];
+        var noOfImageFilesCurrentlyDownloading = 0;
+        // No of images that are not there in the cache and are not currently
+        // being downloaded
+        var noOfImagesNeitherInCacheNorDownloading = 0;
+
+        var state = _states.getState(stateName);
+        imageFilenamesInState = (
+          ExtractImageFilenamesFromStateService
+            .getImageFilenamesInState(state));
+        imageFilenamesInState.forEach(function(filename) {
+          if (!AssetsBackendApiService.isCached(filename) &&
+            (_filenamesOfImageCurrentlyDownloading.indexOf(filename) === -1)) {
+            noOfImagesNeitherInCacheNorDownloading += 1;
+          }
+          if (_filenamesOfImageCurrentlyDownloading.indexOf(filename) !== -1) {
+            noOfImageFilesCurrentlyDownloading += 1;
+          }
+        });
+        if (noOfImagesNeitherInCacheNorDownloading &&
+          noOfImageFilesCurrentlyDownloading <= 1) {
+          _cancelPreloading();
+          _kickOffImagePreloader(stateName);
+        }
+      }
+    };
+
+    return {
+      init: _init,
+      kickOffImagePreloader: _kickOffImagePreloader,
+      onStateChange: _onStateChange,
+      isLoadingImageFile: function(filename) {
+        return _filenamesOfImageCurrentlyDownloading.indexOf(filename) !== -1;
+      },
+      getFilenamesOfImageCurrentlyDownloading: function() {
+        return _filenamesOfImageCurrentlyDownloading;
+      },
+      getImageUrl: function(filename) {
+        return $q(function(resolve, reject){
+          if (AssetsBackendApiService.isCached(filename)) {
+            _getUrlUsingFileInCache(filename, resolve, reject);
```

Comment:
You call `_getUrlUsingFileInCache ` with 3 arguments, but `_getUrlUsingFileInCache ` has onlz 2 defined.

Reply:
Done!